### PR TITLE
New split

### PR DESCRIPTION
**Pull request by reactormonk**
It's about 5x faster.

### CODE REVIEW

**Comment by reactormonk** [comment]
I have to say I have no idea what this breaks. Anyone got a testcase for the corner cases listed in the comments?

File: lib/pure/strutils.nim [hunk 1]
```diff
@@ -241,14 +241,12 @@ iterator split*(s: string, seps: set[char] = Whitespace): string =
...
-  var last = 0
+  var start = 0
   assert(not ('\0' in seps))
-  while last < len(s):
-    while s[last] in seps: inc(last)
-    var first = last
-    while last < len(s) and s[last] notin seps: inc(last) # BUGFIX!
-    if first <= last-1:
-      yield substr(s, first, last-1)
+  for i in 0..len(s):
```
Comment (Araq) [inline lib/pure/strutils.nim]: <len(s)  ? Or is that a subtle feature that you read the terminating zero?

Comment (reactormonk) [inline lib/pure/strutils.nim]: Could you give me an example where that makes a difference?

Comment (Araq) [inline lib/pure/strutils.nim]: No, but reading the \0 is bad style, unless logic requires it.

**Comment by Araq** [comment]
Shouldn't  split("foo", ' ')  produce "foo" instead of nothing?

**Comment by reactormonk** [comment]
`invalid character constant`

**Comment by Araq** [comment]
What?

**Comment by reactormonk** [comment]
Solved, test shows it works as you'd expect it.

**Comment by Araq** [comment]
Well you didn't tell us why it is faster. The algorithms are similar enough that there should not be a difference.

**Comment by reactormonk** [comment]
Couldn't reproduce the benchmark, so I switched to the old algorithm. However, this test fails:

    assert(split(";;this;is;an;;example;;;", ';') == @["this", "is", "an", "example"])

It returns

    @[, , this, is, an, , example, , , ]

**Comment by dom96** [comment]
I'm going to close this. Please create a new PR with code to add those test cases. The test which you think fails doesn't actually fail, it gives the correct output.

**Comment by Araq** [comment]
That's according to split's docs though.

**Comment by reactormonk** [comment]
So @Araq, which one is correct?

**Pull request closed without merging**